library/core/src/iter/traits/iterator.rs

```diff
@@ -500,8 +500,10 @@ pub trait Iterator {
     /// In other words, it zips two iterators together, into a single one.
     ///
     /// If either iterator returns [`None`], [`next`] from the zipped iterator
-    /// will return [`None`]. If the first iterator returns [`None`], `zip` will
-    /// short-circuit and `next` will not be called on the second iterator.
+    /// will return [`None`], in other words zip is short-circuiting as long as one
+    /// of the iterators is finite.
```

Comment:
I wouldn't call this short circuiting, rather it's just the fact that iterators are lazy. Nobody would expect a call to `Zip<A, B>::next` to do multiple calls of `A::next` or `B::next` in a loop. The short-circuiting in the original documentation was, as far as I can tell, referring to how the somewhat logical condition that “If either iterator returns `None`, `next` from the zipped iterator will return `None`” will be evaluated. Taking the “either iterator returns `None`” to basically be a logical disjunction “the first iterator returns `None` *OR* the second iterator returns `None`”, _short-circuiting_ this OR-statement means that the second condition is not evaluated if the first is already true.

Reply:
Interesting, it appears that rust defines "short-circuiting" on iterators differently than [java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#StreamOps) does, which is a big influence for me.

Reply:
Let's avoid the term 'short-circuiting'. In Rust we use that term for e.g. the `&&` and `||` operators which don't evaluate the right hand side if the left hand side already provides the answer. Since `zip` also has a left hand side and right hand side, it can get confusing.